language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionRouter.java

```diff
@@ -99,25 +99,29 @@
                         .isWithinRange(ctx.cursorPosition(), CommonUtil.toRange(diag.location().lineRange()))
                 )
                 .forEach(diagnostic -> {
-                    DiagBasedPositionDetails positionDetails = computePositionDetails(syntaxTree, diagnostic, ctx);
-                    codeActionProvidersHolder.getActiveDiagnosticsBasedProviders(ctx).forEach(provider -> {
-                        try {
-                            // Check whether the code action request has been cancelled
-                            // in order to avoid unnecessary calculations
-                            ctx.checkCancelled();
-                            
-                            List<CodeAction> codeActionsOut = provider.getDiagBasedCodeActions(diagnostic,
-                                                                                               positionDetails, ctx);
-                            if (codeActionsOut != null) {
-                                codeActionsOut.forEach(codeAction ->
-                                        TelemetryUtil.addReportFeatureUsageCommandToCodeAction(codeAction, provider));
-                                codeActions.addAll(codeActionsOut);
+                        DiagBasedPositionDetails positionDetails = computePositionDetails(syntaxTree, diagnostic, ctx);
+                        codeActionProvidersHolder.getActiveDiagnosticsBasedProviders(ctx)
+                                .stream().filter(provider -> provider.validate(diagnostic, positionDetails, ctx))
+                                .forEach(provider -> {
+                            try {
+                                // Check whether the code action request has been cancelled
+                                // in order to avoid unnecessary calculations
+                                ctx.checkCancelled();
+
+                                List<CodeAction> codeActionsOut = provider.getDiagBasedCodeActions(diagnostic,
+                                        positionDetails, ctx);
+                                if (codeActionsOut != null) {
```

Comment:
This has been in the previous implementation as well. Check whether why we need the null check here. Each provider should return an empty list in case of error scenarios.

Reply:
Checked the implementation in each code action provider. There are no methods returning null. Hence removed the null check.